20240917-BR-resiliency-error-code-retries.md

```diff
@@ -0,0 +1,150 @@
+https://github.com/dapr/dapr/issues/7697
+
+PR:
+https://github.com/dapr/dapr/pull/7132
+
+Docs:
+https://github.com/dapr/docs/issues/4254
+https://github.com/dapr/docs/issues/3859
+
+## Expectations and alternatives
+
+* What is in scope for this proposal?
+  - HTTP and gRPC Service Invocation, direct and proxied
+  - Bindings
+  - Pub/Sub
+
+## Implementation Details
+
+### Design
+
+Add a new object field to the `retries` policy Spec to allow the user to specify the status codes that should be retried.
+Separate fields for HTTP and gRPC. The new fields should be optional and will default to the existing behavior, which is to retry on all errors.
+
+### Example 1:
+In this example, the retry policy will retry **_only_** on HTTP 500 and HTTP status code range 502-504 (inclusive) and gRPC status code range 2-4 (inclusive).
+The rest of the status codes will not be retried.
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Resiliency
+metadata:
+  name: myresiliency
+scopes:
+  - app1
+spec:
+  policies:
+    retries:
+      pubsubRetry:
+        policy: constant
+        duration: 5s
+        maxRetries: 10
+        matching:
+          httpStatusCodes: "500,502-504"
+          gRPCStatusCodes: "2-4"
+```
+
+### Example 2:
+In this example, the retry policy will retry **_only_** on gRPC status code range 1-15 (inclusive).
+However, this policy will not apply to the HTTP status codes, and they will be retried according to the default behavior, which is to retry on all errors.
+
+```yaml
+apiVersion: dapr.io/v1alpha1
+kind: Resiliency
+metadata:
+  name: myresiliency
+scopes:
+  - app1
+spec:
+  policies:
+    retries:
+      pubsubRetry:
+        policy: constant
+        duration: 5s
+        maxRetries: 10
+        matching:
+          gRPCStatusCodes: "1-15"
+```
+
+### Acceptable Values
+The acceptable values are the same as the ones defined in the [HTTP Status Codes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status) and [gRPC Status Codes](https://grpc.io/docs/guides/status-codes/) documentation.
+
+- HTTP: from 100 to 599
+- gRPC: from 1 to 16
+
+### Setting Format
+Both the `httpStatusCodes` and `gRPCStatusCodes` fields are of type string and optional and can be set to a comma-separated list of status codes and/or ranges of status codes.
+The range must be in the format `<start>-<end>` (inclusive). Having more than one dash in the range is not allowed.
+
+### CRD Validation
+
+Both field values should be validated using Common Expression Language [CEL](https://kubernetes.io/docs/reference/using-api/cel/)
+In addition, see Kubebuilder documentation for [CRD Validation](https://book.kubebuilder.io/reference/markers/crd-validation)
+
+### Parsing the configuration
+
+The configuration values will be first parsed as comma-separated lists.
+Each entry in the list will be then parsed as a single status code or a range of status codes.
+Invalid entries will be logged and the Dapr runtime will fail to start.
```

Comment:
While adding this test, I found that this is not how it's happening today.
Any error in loading resiliency will be logged, but the `daprd` continues to run.
https://github.com/dapr/dapr/blob/8b7020a5b21419c564cc14bcbc4d4c1e019b3313/pkg/resiliency/resiliency.go#L294-L297
Changing this behavior will affect `daprd` startup and can produce unexpected results.
Have a few options on how to handle this:
1. leave as today, log and continue
2. intentional `Fatal` at the specific place when failing to parse the fields [here](https://github.com/dapr/dapr/blob/4c3efd73cdcdeddf1eb457016af92cf67d466966/pkg/resiliency/resiliency.go#L427)
3. introduce a special `error` type to fail `daprd` only on this case and skip on the rest
Thoughts?

Reply:
It’s unfortunate that today parsing a bad manifest will be completely hidden by the user until something goes wrong. Resiliency policy is certainly not the only offender on this, however it’s particularly bad for resiliency as the manifestation of non-consumed policy isn’t necessarily distinguishable from default behaviour unless you’re really looking for it.

Though not happy about it, I’m in favour of continuing the same behaviour as we have today, as in, ignore the policy if it is invalid, and address this issue separately.

Long term, we need to add proper validation at every API ingestion level (i.e. at kube API, validating webhook, and Dapr parading), as well as status reporting for all resources. We have a [components status reporting](https://github.com/dapr/proposals/pull/54) proposal already which likely has the most impact to users, but it should be expanded to all resources.